mypy errors for prange

<!--

Thanks for opening an issue! To help the Numba team handle your information
efficiently, please first ensure that there is no other issue present that
already describes the issue you have
(search at https://github.com/numba/numba/issues?&q=is%3Aissue).

-->

## Reporting a bug

<!--

Before submitting a bug report please ensure that you can check off these boxes:

-->

- [x] I have tried using the latest released version of Numba (most recent is
  visible in the release notes
  (https://numba.readthedocs.io/en/stable/release-notes-overview.html).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

<!--

Please include details of the bug here, including, if applicable, what you
expected to happen!

-->
numba 0.66.0 introduces numba/py.typed, which causes extra type checking, and causes mypy to complain about prange (I think because the Python proxy class that gets lowered isn't iterable, and mypy doesn't know about jitting).

To reproduce:
```python
#!/usr/bin/env python3

import numpy as np
from numba import njit, prange

@njit
def add(x, y) -> None:
    for i in prange(len(x)):
        x[i] += y[i]
```

mypy output:
```
prange_typing.py:8: error: "prange" has no attribute "__iter__" (not iterable)  [attr-defined]
```

--- a/maint/stubtest/allowlist.txt
+++ b/maint/stubtest/allowlist.txt
@@ -24,7 +24,16 @@ numba\.core\.utils\..*
 numba\.cpython\..*
 numba\.cuda\..*
 numba\.experimental\..*
-numba\.misc\..*
+numba\.misc\.POST
+numba\.misc\.appdirs\..*
+numba\.misc\.coverage_support\..*
+numba\.misc\.dump_style.*
+numba\.misc\.gdb_print_extension
+numba\.misc\.help\..*
+numba\.misc\.numba_entry
+numba\.misc\.numba_gdbinfo
+numba\.misc\.numba_sysinfo
+numba\.misc\.timsort
 numba\.mviewbuf
 numba\.np\.(arraymath|arrayobj|linalg|npyimpl|random|types|ufunc_db|unsafe)\..*
 numba\.np\.extensions

--- a/numba/misc/special.pyi
+++ b/numba/misc/special.pyi
@@ -1,0 +1,34 @@
+from builtins import range as prange
+from typing import TypeVar
+
+from _typeshed import Incomplete
+from numpy import ndindex as pndindex
+
+from numba.core.typing.asnumbatype import as_numba_type
+from numba.core.typing.typeof import typeof
+
+__all__ = [
+    "typeof",
+    "as_numba_type",
+    "prange",
+    "pndindex",
+    "gdb",
+    "gdb_breakpoint",
+    "gdb_init",
+    "literally",
+    "literal_unroll",
+]
+
+###
+
+_T = TypeVar("_T")
+
+###
+
+def gdb(*args: Incomplete) -> None: ...
+def gdb_breakpoint() -> None: ...
+def gdb_init(*args: Incomplete) -> None: ...
+
+#
+def literally(obj: _T) -> _T: ...
+def literal_unroll(container: _T) -> _T: ...

